### PR TITLE
WIP: Stunnel sidecar container

### PR DESCRIPTION
**Pull request by kyguy**
Haven't been able to get the sidecar working properly just yet

The stunnel process has been running to completion instead of staying in the background. I just realized I am installing outdated version of stunnel(4.56) instead of (5.48) and believe that is the source of the issue. In the meantime, I'm having the container execute a different (unending)process so I can debug the container as it runs.

Apologies for the delay in getting this working ( I understand this blocks the cert work ), feel free to have a go at it; regardless, I'll be back hacking this later

### CODE REVIEW

**Comment by kyguy** [comment]
Switched to fedora 27 image for an easy stunnel upgrade(just for testing) but running into the same issues as before. When I declare "CMD ["/opt/kafka/stunnel_run.sh"]" in the docker image, it appears that "stunnel /tmp/stunnel.conf" is never called( or fails) causing the container to be terminated

**Comment by tombentley** [comment]
Have you tried with `foreground = yes` in the config? By default it will daemonize, which is probably what's causing the container to exit.

**Comment by kyguy** [comment]
Tom you are the best! I was incorrectly defining "foreground = yes" in the config causing it to fail. I wouldn't have noticed this if you hadn't pointed it out. Its working now, thanks!

File: docker-images/stunnel-zookeeper/scripts/stunnel_config_generator.sh [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+		# Not necessarily the way to create pem file, might have to use openssl
+		cat ${KEY_AND_CERT}.crt ${KEY_AND_CERT}.key > ${PEM}
+		chmod 640 ${PEM}
+
+		# Stunnel client configuration
+		cat <<-EOF
+		[${PEER})]
+		client=yes
+		cert=${PEM}
```
Comment (ppatierno) [inline docker-images/stunnel-zookeeper/scripts/stunnel_config_generator.sh]: I think that it should be the "internal-ca" certificate so you should use `CAfile` parameter and not `cert`.
In this client configuration, the stunnel is connecting to a remote peer which will send the server certificate signed by the "internal-ca" cert; for this reason the stunnel client has to verify the signed certificate (received by the peer during the TLS handshake) using such CA cert.

File: docker-images/stunnel-zookeeper/scripts/stunnel_config_generator.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+		accept=127.0.0.1:$(expr $port \* 10 + $NODE)
+		connect=${PEER}.${BASE_FQDN}:$port
+
+		EOF
+		let NODE=NODE+1
+  done
+	# Zookeeper port where stunnel forwards recieved traffic
+	CONNECTOR_PORT=$(expr $port \* 10)
```
Comment (ppatierno) [inline docker-images/stunnel-zookeeper/scripts/stunnel_config_generator.sh]: Using the 28880 and 38880 ports (which is what I set in the CO as well) shifts the listeners ports of stunnel to 28881 for zookeeper-0, 28882 for zookeeper-1 and so on. If we want to stick with the rule that port = 2888 * 10 + index (where index starts from 0), we should change the 28880 and 38880 ports. What do you think about something like 2999 and 3999 ? @kyguy @tombentley @scholzj

**Comment by ppatierno** [comment]
@kyguy I pushed a couple of commit for having it working. We need to clean up few things btw before merging with my `tls-zookeeper-stunnel` and then opening a PR against the `master`.

**Pull request merged**